src/_pytest/python.py

```diff
@@ -1220,32 +1278,41 @@ def _idvalset(idx, parameterset, argnames, idfn, ids, item, config):
         ]
         return "-".join(this_id)
     else:
-        return _ascii_escaped_by_config(ids[idx], config)
-
-
-def idmaker(argnames, parametersets, idfn=None, ids=None, config=None, item=None):
-    ids = [
+        id = ids[idx]
+        assert id is not None  # Checked above.
```

Comment:
I wonder if this could also have a type comment so that it can be detected/removed when mypy recognizes it?

Reply:
Not that I know of.

This can be rewritten to avoid the need for the assert:

```diff
diff --git a/src/_pytest/python.py b/src/_pytest/python.py
index e4d7cbf86..5edc89322 100644
--- a/src/_pytest/python.py
+++ b/src/_pytest/python.py
@@ -1271,15 +1271,14 @@ def _idvalset(
 ):
     if parameterset.id is not None:
         return parameterset.id
-    if ids is None or (idx >= len(ids) or ids[idx] is None):
+    id = None if ids is None or idx >= len(ids) else ids[idx]
+    if id is None:
         this_id = [
             _idval(val, argname, idx, idfn, item=item, config=config)
             for val, argname in zip(parameterset.values, argnames)
         ]
         return "-".join(this_id)
     else:
-        id = ids[idx]
-        assert id is not None  # Checked above.
         return _ascii_escaped_by_config(id, config)
```

Let's me know if you think it's better.

Reply:
This seems to be a bit better also (avoiding a comment/explanation also).